From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: David Llewellyn-Jones <david.llewellyn-jones@jolla.com>
Date: Thu, 21 Oct 2021 12:35:53 +0000
Subject: [PATCH 48/67] Allow LoginManagerPrompter to find its window.
 JB#55760, OMP#JOLLA-418

This patch blocks loading of gecko's LoginManagerAuthPrompter.jsm so
that the embedlite-components version can be used instead.

It also patches the nsILoginManagerPrompter interface to allow a
reference to the window to be passed through, to allow the embedlite
component to understand its context.

Finally it patches ParentChannelListener to pass the correct window
object through to the nsILoginManagerAuthPrompter component.
---
 .../protocol/http/ParentChannelListener.cpp    |  2 +-
 .../passwordmgr/LoginManagerParent.sys.mjs     |  4 ++--
 toolkit/components/passwordmgr/components.conf | 18 ------------------
 .../passwordmgr/nsILoginManagerPrompter.idl    |  6 +++---
 4 files changed, 6 insertions(+), 24 deletions(-)

diff --git a/netwerk/protocol/http/ParentChannelListener.cpp b/netwerk/protocol/http/ParentChannelListener.cpp
index 34262f6d9792..d71386b58b50 100644
--- a/netwerk/protocol/http/ParentChannelListener.cpp
+++ b/netwerk/protocol/http/ParentChannelListener.cpp
@@ -241,7 +241,7 @@ ParentChannelListener::GetAuthPrompt(uint32_t aPromptReason, const nsIID& iid,
 
   nsCOMPtr<nsPIDOMWindowOuter> window;
   RefPtr<dom::Element> frame = mBrowsingContext->Top()->GetEmbedderElement();
-  if (frame) window = frame->OwnerDoc()->GetWindow();
+  window = frame ? frame->OwnerDoc()->GetWindow() : mBrowsingContext->GetDOMWindow();
 
   // Get an auth prompter for our window so that the parenting
   // of the dialogs works as it should when using tabs.
diff --git a/toolkit/components/passwordmgr/LoginManagerParent.sys.mjs b/toolkit/components/passwordmgr/LoginManagerParent.sys.mjs
index dcf770b9d136..0bc128004163 100644
--- a/toolkit/components/passwordmgr/LoginManagerParent.sys.mjs
+++ b/toolkit/components/passwordmgr/LoginManagerParent.sys.mjs
@@ -1084,7 +1084,7 @@ export class LoginManagerParent extends JSWindowActorParent {
       existingLogin = this.#findSameLogin(logins, formLogin);
     }
 
-    const promptBrowser = lazy.LoginHelper.getBrowserForPrompt(browser);
+    const promptBrowser = browsingContext.window;
     const prompter = this._getPrompter(browser);
 
     if (!canMatchExistingLogin) {
@@ -1420,7 +1420,7 @@ export class LoginManagerParent extends JSWindowActorParent {
     }
 
     const prompter = this._getPrompter(browser);
-    const promptBrowser = lazy.LoginHelper.getBrowserForPrompt(browser);
+    const promptBrowser = browsingContext.window;
 
     if (existingLogin) {
       // Show a change doorhanger to allow modifying an already-saved login
diff --git a/toolkit/components/passwordmgr/components.conf b/toolkit/components/passwordmgr/components.conf
index 9be23b98f558..00d5dd83b3b4 100644
--- a/toolkit/components/passwordmgr/components.conf
+++ b/toolkit/components/passwordmgr/components.conf
@@ -13,24 +13,12 @@ Classes = [
         'esModule': 'resource://gre/modules/LoginManager.sys.mjs',
         'constructor': 'LoginManager',
     },
-    {
-        'cid': '{749e62f4-60ae-4569-a8a2-de78b649660e}',
-        'contract_ids': ['@mozilla.org/passwordmanager/authpromptfactory;1'],
-        'esModule': 'resource://gre/modules/LoginManagerAuthPrompter.sys.mjs',
-        'constructor': 'LoginManagerAuthPromptFactory',
-    },
     {
         'cid': '{2bdac17c-53f1-4896-a521-682ccdeef3a8}',
         'contract_ids': ['@mozilla.org/login-manager/autocompletesearch;1'],
         'esModule': 'resource://gre/modules/LoginAutoComplete.sys.mjs',
         'constructor': 'LoginAutoComplete',
     },
-    {
-        'cid': '{8aa66d77-1bbb-45a6-991e-b8f47751c291}',
-        'contract_ids': ['@mozilla.org/login-manager/authprompter;1'],
-        'esModule': 'resource://gre/modules/LoginManagerAuthPrompter.sys.mjs',
-        'constructor': 'LoginManagerAuthPrompter',
-    },
     {
         'cid': '{0f2f347c-1e4f-40cc-8efd-792dea70a85e}',
         'contract_ids': ['@mozilla.org/login-manager/loginInfo;1'],
@@ -68,12 +56,6 @@ else:
             'esModule': 'resource://gre/modules/storage-json.sys.mjs',
             'constructor': 'LoginManagerStorage_json',
         },
-        {
-            'cid': '{c47ff942-9678-44a5-bc9b-05e0d676c79c}',
-            'contract_ids': ['@mozilla.org/login-manager/prompter;1'],
-            'esModule': 'resource://gre/modules/LoginManagerPrompter.sys.mjs',
-            'constructor': 'LoginManagerPrompter',
-        },
         {
             'cid': '{dc185a77-ba88-4caa-8f16-465253f7599a}',
             'contract_ids': [
diff --git a/toolkit/components/passwordmgr/nsILoginManagerPrompter.idl b/toolkit/components/passwordmgr/nsILoginManagerPrompter.idl
index 4f244258c0f5..8b162b92b29c 100644
--- a/toolkit/components/passwordmgr/nsILoginManagerPrompter.idl
+++ b/toolkit/components/passwordmgr/nsILoginManagerPrompter.idl
@@ -34,7 +34,7 @@ interface nsILoginManagerPrompter : nsISupports {
    *        a username or password.  Has two properties, 'usernames' and 'passwords'.
    */
   nsIPromptInstance promptToSavePassword(
-          in Element aBrowser,
+          in nsIDOMWindow aBrowser,
           in nsILoginInfo aLogin,
           [optional] in boolean dismissed,
           [optional] in boolean notifySaved,
@@ -64,7 +64,7 @@ interface nsILoginManagerPrompter : nsISupports {
    *        a username or password.  Has two properties, 'usernames' and 'passwords'.
    */
   nsIPromptInstance promptToChangePassword(
-          in Element aBrowser,
+          in nsIDOMWindow aBrowser,
           in nsILoginInfo aOldLogin,
           in nsILoginInfo aNewLogin,
           [optional] in boolean dismissed,
@@ -92,7 +92,7 @@ interface nsILoginManagerPrompter : nsISupports {
    *       is called.
    */
   nsIPromptInstance promptToChangePasswordWithUsernames(
-          in Element aBrowser,
+          in nsIDOMWindow aBrowser,
           in Array<nsILoginInfo> logins,
           in nsILoginInfo aNewLogin);
 };
